### DISCUSSION TITLE
Config File

### DISCUSSION THREAD

**Discussion by karreglover** [helgeerbe/picframe#212] [Q&A] [answered]
Hi, I am trying to set up a picframe on Raspberry Pi Zero using the instructions found on:  https://www.thedigitalpictureframe.com/how-to-add-crossfading-slide-transitions-to-your-digital-picture-frame-using-pi3d/

Unfortunately I have very little experience with coding, and when I get to this step: 
pi@raspberrypiglover:~ $ picframe ~/picframe_data/config/configuration.yaml

I get the following message:

INFO:start.py:starting ['/home/pi/.local/bin/picframe', '/home/pi/picframe_data/config/configuration.yaml']
INFO:model.Model:Open config file: /home/pi/picframe_data/config/configuration.yaml:
Traceback (most recent call last):
  File "/home/pi/.local/bin/picframe", line 8, in <module>
    sys.exit(main())
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/start.py", line 130, in main
    m = model.Model(args.configfile)
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/model.py", line 173, in __init__
    self.__image_cache = image_cache.ImageCache(self.__pic_dir,
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/image_cache.py", line 41, in __init__
    self.__update_schema(3)
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/image_cache.py", line 319, in __update_schema
    self.__db.execute("ALTER TABLE folder ADD COLUMN missing INTEGER DEFAULT 0 NOT NULL")
sqlite3.OperationalError: duplicate column name: missing

Any clue where I went wrong?  TIA.

**Comment by helgeerbe** [comment]
Hm, I can't tell what went wrong. What happens if you delete the database `~/picframe_data/data/pictureframe.db3`? If you start picframe database and cache should be recreated.

**Comment by sapnho** [comment]
Are you working through a Terminal or directly on your Pi with an attached keyboard?

**Reply by karreglover** [reply]
I am working through a Terminal.

I deleted the database, reinstalled, and got this message:

pi@raspberrypiglover:~ $ pip3 install picframe
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Requirement already satisfied: picframe in ./.local/lib/python3.9/site-packages (2021.11.4)
Requirement already satisfied: ExifRead in ./.local/lib/python3.9/site-packages (from picframe) (2.3.2)
Requirement already satisfied: PyYAML in /usr/lib/python3/dist-packages (from picframe) (5.3.1)
Requirement already satisfied: IPTCInfo3 in ./.local/lib/python3.9/site-packages (from picframe) (2.1.4)
Requirement already satisfied: numpy in /usr/lib/python3/dist-packages (from picframe) (1.19.5)
Requirement already satisfied: pi3d>=2.49 in ./.local/lib/python3.9/site-packages (from picframe) (2.49)
Requirement already satisfied: ninepatch in ./.local/lib/python3.9/site-packages (from picframe) (0.1.20)
Requirement already satisfied: Pillow in /usr/lib/python3/dist-packages (from picframe) (8.1.2)
Requirement already satisfied: paho-mqtt in ./.local/lib/python3.9/site-packages (from picframe) (1.6.1)
pi@raspberrypiglover:~ $ picframe ~/picframe_data/config/configuration.yaml
INFO:start.py:starting ['/home/pi/.local/bin/picframe', '/home/pi/picframe_data/config/configuration.yaml']
INFO:model.Model:Open config file: /home/pi/picframe_data/config/configuration.yaml:
Traceback (most recent call last):
  File "/home/pi/.local/bin/picframe", line 8, in <module>
    sys.exit(main())
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/start.py", line 136, in main
    c.start()
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/controller.py", line 317, in start
    self.__viewer.slideshow_start()
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/viewer_display.py", line 416, in slideshow_start
    self.__display = pi3d.Display.create(x=self.__display_x, y=self.__display_y,
  File "/home/pi/.local/lib/python3.9/site-packages/pi3d/Display.py", line 569, in create
    display.opengl.create_display(x, y, w, h, depth=depth, samples=samples, layer=layer,
  File "/home/pi/.local/lib/python3.9/site-packages/pi3d/util/DisplayOpenGL.py", line 122, in create_display
    assert self.context != EGL_NO_CONTEXT and self.context is not None
AssertionError
^CException ignored in: <module 'threading' from '/usr/lib/python3.9/threading.py'>
Traceback (most recent call last):
  File "/usr/lib/python3.9/threading.py", line 1428, in _shutdown
    lock.acquire()
KeyboardInterrupt:
pi@raspberrypiglover:~ $

**Comment by helgeerbe** [comment]
This looks quite OK now. No more database problem.
- Which OS are you using? buster or bullseye? `cat /etc/os-release`
- Do you boot to desktop or command prompt? If you boot to command prompt. Try `export DISPLAY=:0` before starting picframe. This tells X system on which display it has to show picframe.

**Reply by karreglover** [reply]
I am using bullseye and it boots to the desktop.  

This is what I tried this time and what I got:

pi@raspberrypiglover:~ $ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
NAME="Raspbian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"
pi@raspberrypiglover:~ $ export DISPLAY=:0
pi@raspberrypiglover:~ $ picframe
INFO:start.py:starting ['/home/pi/.local/bin/picframe']
INFO:model.Model:Open config file: /home/pi/picframe_data/config/configuration.yaml:
Traceback (most recent call last):
  File "/home/pi/.local/bin/picframe", line 8, in <module>
    sys.exit(main())
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/start.py", line 132, in main
    m = model.Model()
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/model.py", line 173, in __init__
    self.__image_cache = image_cache.ImageCache(self.__pic_dir,
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/image_cache.py", line 41, in __init__
    self.__update_schema(3)
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/image_cache.py", line 319, in __update_schema
    self.__db.execute("ALTER TABLE folder ADD COLUMN missing INTEGER DEFAULT 0 NOT NULL")
sqlite3.OperationalError: duplicate column name: missing

**Comment by jgodfrey** [comment]
Hmmm... Seems like you're back to the database schema migration error again... What happens if you remove the cache database file and start the app again (no need to reinstall)?

**Comment by karreglover** [comment]
No luck.  Still get this error:

***@***.***:~ $ picframe
INFO:start.py:starting ['/home/pi/.local/bin/picframe']
INFO:model.Model:Open config file:
/home/pi/picframe_data/config/configuration.yaml:
Traceback (most recent call last):
  File "/home/pi/.local/bin/picframe", line 8, in <module>
    sys.exit(main())
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/start.py",
line 132, in main
    m = model.Model()
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/model.py",
line 173, in __init__
    self.__image_cache = image_cache.ImageCache(self.__pic_dir,
  File
"/home/pi/.local/lib/python3.9/site-packages/picframe/image_cache.py", line
41, in __init__
    self.__update_schema(3)
  File
"/home/pi/.local/lib/python3.9/site-packages/picframe/image_cache.py", line
319, in __update_schema
    self.__db.execute("ALTER TABLE folder ADD COLUMN missing INTEGER
DEFAULT 0 NOT NULL")
sqlite3.OperationalError: duplicate column name: missing

**Comment by jgodfrey** [comment]
Is it possible to make the broken db file available for examination? It's trying to auto-migrate your db schema (as it should), but somehow the schema version marker is out of sync with the actual schema. Looking at the code, I don't see how that can happen, but it obviously is.

**Comment by karreglover** [comment]
Hopefully this attachment worked.

**Comment by jgodfrey** [comment]
Hmmm... I don't see any obvious attachment here.

**Reply by karreglover** [reply]
It's not letting me attach the file.  Says that type of file is not supported.

**Comment by jgodfrey** [comment]
Maybe put the db file in a ZIP before attaching it?

Edit - or, perhaps just rename the file with an extension that is supported here (I'm not sure whether that'll work or not).

**Reply by karreglover** [reply]
[pictureframe.zip](https://github.com/helgeerbe/picframe/files/7829696/pictureframe.zip)

I think that worked!

**Comment by jgodfrey** [comment]
Just took a quick look. The schema of the db has been properly upgraded to v3 (as intended), but the schema version number hasn't been recorded for some reason. Since there's no schema version number in the db, the migration code assumes it needs to be migrated again - which leads to the error. I do see that we don't explicitly commit the migration changes in the code, which (coupled with some other unexpected, downstream error) could be the root cause here.

Anyway, I've added the schema version tag to the attached db. If you replace your existing db with the attached, I don't think you'll see the db error again. That said, I'll need to get to the bottom of whatever is causing the underlying migration  issue.

[pictureframe.zip](https://github.com/helgeerbe/picframe/files/7830053/pictureframe.zip)

**Comment by jgodfrey** [comment]
I do think the missing `commit()` call in the migration code is the most likely candidate here. I'll try to merge a fix for that this evening.

**Comment by jgodfrey** [comment]
One more comment here...

I don't think the DB issue is the real problem here (although it really is a bug). When the app first starts up, if the db doesn't yet exist (first startup), it's created (and its schema is migrated to the latest version).  It's not until much later in the execution path that we actually begin to add (and commit) image data to the db. 

Between the time the initial db is created and data is actually inserted in it, if anything goes wrong (so, the app crashes), the db is left in a state where the schema is actually correct, but the version number hasn't been recorded yet. At that point, if you RESTART the app, the migration code sees there is no schema version number and attempts to migrate the db again. Since it's already migrated, that leads to the db errors reported above.

That issue will be fixed when I merge the above-mentioned `commit()` changes soon. While that should fix the DB error on successive starts, I don't think it has anything to do with the original problem. Looking over the crash dumps posted above, the underlying problem seems to be this:

```
File "/home/pi/.local/lib/python3.9/site-packages/pi3d/util/DisplayOpenGL.py", line 122, in create_display
assert self.context != EGL_NO_CONTEXT and self.context is not None
AssertionError
^CException ignored in: <module 'threading' from '/usr/lib/python3.9/threading.py'>
Traceback (most recent call last):
File "/usr/lib/python3.9/threading.py", line 1428, in _shutdown
```

That's an issue I can't really speak to, but I expect that @paddywwoof or @helgeerbe might have some input.

**Reply by karreglover** [reply]
That seemed to fix the one error, but now I'm back at this:

pi@raspberrypiglover:~ $ picframe ~/picframe_data/config/configuration.yaml
INFO:start.py:starting ['/home/pi/.local/bin/picframe', '/home/pi/picframe_data/config/configuration.yaml']
INFO:model.Model:Open config file: /home/pi/picframe_data/config/configuration.yaml:
Traceback (most recent call last):
  File "/home/pi/.local/bin/picframe", line 8, in <module>
    sys.exit(main())
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/start.py", line 136, in main
    c.start()
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/controller.py", line 317, in start
    self.__viewer.slideshow_start()
  File "/home/pi/.local/lib/python3.9/site-packages/picframe/viewer_display.py", line 416, in slideshow_start
    self.__display = pi3d.Display.create(x=self.__display_x, y=self.__display_y,
  File "/home/pi/.local/lib/python3.9/site-packages/pi3d/Display.py", line 569, in create
    display.opengl.create_display(x, y, w, h, depth=depth, samples=samples, layer=layer,
  File "/home/pi/.local/lib/python3.9/site-packages/pi3d/util/DisplayOpenGL.py", line 122, in create_display
    assert self.context != EGL_NO_CONTEXT and self.context is not None
AssertionError
^X^CException ignored in: <module 'threading' from '/usr/lib/python3.9/threading.py'>
Traceback (most recent call last):
  File "/usr/lib/python3.9/threading.py", line 1428, in _shutdown
    lock.acquire()
KeyboardInterrupt:

**Comment by jgodfrey** [comment]
Committed https://github.com/helgeerbe/picframe/commit/ce7483fd2013c1650681152c5057b5415bfea400 to `dev` to fix the schema commit issue...

**Comment by jgodfrey** [comment]
> That seemed to fix the one error, but now I'm back at this:

Yeah, I guess that's what I would expect (and mentioned above). Let's see if @paddywwoof or @helgeerbe have any input there.

**Comment by paddywwoof** [comment]
Hi, I'm not sure if I've read this correctly but the error looks like the problem with bullseye and the new GL drivers. I think Wolfgang mentions on his blog the availability of the buster SD image supported by the Pi foundation. I would stick to the tried and tested OS for the moment and see if that cures the problem.

**Comment by helgeerbe** [comment]
If you want to keep bullseye you can try this: https://github.com/helgeerbe/picframe/issues/195#issuecomment-982999227

Otherwise use the legacy buster.

**Comment by karreglover** [comment]
I appreciate all the help.  I've gotten buster OS installed now and I can
see the "No Pictures Selected Image) when I start picframe.  However, I am
still getting an error message:

***@***.***:~ $ picframe
INFO:start.py:starting ['/home/pi/.local/bin/picframe']
INFO:model.Model:Open config file:
/home/pi/picframe_data/config/configuration.yaml:
ERROR:model.Model:Can't parse yaml config file:
/home/pi/picframe_data/config/configuration.yaml: expected '<document
start>', but found '<block mapping start>'
  in "/home/pi/picframe_data/config/configuration.yaml", line 41, column 1
ERROR:model.Model:error trying to set locale to en_US.utf8

Any clues?

**Reply by jgodfrey** [reply]
Looks like you have a broken configuration file. I assume you've made an edit that has somehow broken the syntax of the file. If you can't find it, feel free to post the file for some assistance.

**Comment by karreglover** [comment]
This is my file:

 viewer:
  blur_amount: 12                         # default=12,  larger values than
12 will increase processing load quite a bit
  blur_zoom: 1.0                          # default=1.0, must be >= 1.0
which expands the backgorund to just fill the space around the image
  blur_edges: False                       # default=False, use blurred
version of image to fill edges - will override FIT = False
  edge_alpha: 0.5                         # default=0.5, background colour
at edge. 1.0 would show reflection of image
  fps: 20.0                               # default=20.0
  background: [0.2, 0.2, 0.3, 1.0]        # default=[0.2, 0.2, 0.3, 1.0],
RGBA to fill edges when fitting
  blend_type: "blend"                     # default="blend",
choices={"blend", "burn", "bump"}, type of blend the shader can do
  font_file: "/home/pi/picframe_data/data/fonts/NotoSans-Regular.ttf"
  shader: "/home/pi/picframe_data/data/shaders/blend_new"
  show_text_fm: "%b %d, %Y"               # default "%b %d, %Y", format to
show date over the image
  show_text_tm: 20.0                      # default=20.0, time to show text
over image with file name
  show_text_sz: 40                        # default=40, text character size
  show_text: ""                           # default="title caption name
date folder location", show text, include combination of words: title,
caption name, date, location, folder
  text_justify: "L"                       # text justification L, C or R
  fit: False                              # default=False, True => scale
image so all visible and leave 'gaps'
                                          #                False => crop
image so no 'gaps'
  kenburns: False                         # default=False, will set
fit->False and blur_edges->False
  display_x: 0                            # offset from left of screen (can
be negative)
  display_y: 0                            # offset from top of screen (can
be negative)
  display_w: null                         # width of display surface
(null->None will use max returned by hardware)
  display_h: null                         # height of display surface
  use_glx: False                          # default=False. Set to True on
linux with xserver running

  mat_images: 0.01                        # default=0.01, True,
automatically mat all images. False, don't automatically mat any images.
Real value, auto-mat all images with aspect ratio difference > than value
  mat_type: null                          # default=null, A string
containing the mat types to choose from when matting images. It can consist
of any or
                                          # all of 'float float_polaroid
float_color_wrap single_bevel double_bevel double_flat' (null or '' will
use all mat types)
  outer_mat_color: null                   # default=null, Color of the
outer mat as an RGB list. null will auto-select a reasonable color based on
the image.
  inner_mat_color: null                   # default=null, Color of the
inner mat as an RGB list. null will auto-select a reasonable color based on
the image.
  outer_mat_border: 75                    # default=75, Minimum outer mat
border in pixels
  inner_mat_border: 40                    # default=40, Minimum inner mat
border in pixels (for styles that use it)
  outer_mat_use_texture: True             # default=True, True uses a
texture for the outer mat. False creates a solid-color outer mat.
  inner_mat_use_texture: False            # default=False, True uses a
texture for the inner mat. False creates a solid-color inner mat.
  mat_resource_folder: "/home/pi/picframe_data/data/mat" # Folder
containing mat image files

  show_clock: False                       # default=False, True shows clock
overlay. False does not show clock overlay
  clock_justify: "R"                      # default="R", clock
justification L, C, or R
  clock_text_sz: 120                      # default=120, clock character
size
  clock_format: "%I:%M"                   # default="%I:%M", strftime
format for clock string

model:
  pic_dir: "/home/pi/Pictures"                   #
default="/home/pi/Pictures", root folder for images
  deleted_pictures: "/home/pi/DeletedPictures"   # move deleted pictures
here
  follow_links: False                     # default=False, By default,
picframe will not walk down into symbolic links that resolve to
directories. Set follow_links to True to visit directories pointed to by
symlinks, on systems that support them.
  no_files_img: "/home/pi/picframe_data/data/no_pictures.jpg" #
default="PictureFrame2020img.jpg", image to show if none selected
  subdirectory: ""                        # default="", subdir of pic_dir -
can be changed by MQTT"
  recent_n: 7                             # default=7 (days), when
shuffling file change date more recent than this number of days play before
the rest
  reshuffle_num: 1                        # default=1, times through before
reshuffling
  time_delay: 200.0                       # default=200.0, time between
consecutive slide starts - can be changed by MQTT
  fade_time: 10.0                         # default=10.0, change time
during which slides overlap - can be changed by MQTT"
  shuffle: True                           # default=True, shuffle on
reloading image files - can be changed by MQTT"
  sort_cols: 'fname ASC'                  # default='fname ASC' can be any
columns in the table with optional ASC or DESC separated by commas
                                          # fname, last_modified, file_id,
orientation, exif_datetime, f_number,
                                          # exposure_time, iso,
focal_length, make, model, lens, rating,
                                          # latitude, longitude, width,
height, title, caption, tags,
                                          # is_portrait, location
  image_attr: [                           # image attributes send by MQTT,
Keys are taken from exifread library, "PICFRAME GPS" is special to retrieve
GPS lon/lat, "PICFRAME LOCATION" is special to retrieve geo reverse
(load_geoloc hast to be True)
    "PICFRAME GPS",
    "PICFRAME LOCATION",
    "EXIF FNumber",
    "EXIF ExposureTime",
    "EXIF ISOSpeedRatings",
    "EXIF FocalLength",
    "EXIF DateTimeOriginal",
    "Image Model",
    "Image Make",
    "IPTC Caption/Abstract",
    "IPTC Object Name",
    "IPTC Keywords"]
  load_geoloc: False                      # get location information from
open street map NB if you switch this on (recommended)
  geo_key: ***@***.***_changed"     # then you **MUST** change the
geo_key to something unique to you
                                          # i.e. use your email address
  locale: "en_GB.UTF-8"                    # "locale -a" shows the
installed locales which could used
  key_list: [
    ["tourism","amenity","isolated_dwelling"],
    ["suburb","village"],
    ["city","county"],
    ["region","state","province"],
    ["country"]]
  db_file: "/home/pi/picframe_data/data/pictureframe.db3" # database used
by PictureFrame
  portrait_pairs: False
  log_level: "WARNING"                    # default=WARNING, could beDEBUG,
INFO, WARNING, ERROR, CRITICAL
  log_file: ""                            # default="" for debugging set
this to the path to a file. NB logging messages will
                                          # appended indefinitely so don't
forget this. You will need to tidy it up later
  use_kbd: False                          # default=False, just for debug
or console start. Crashing when started with systemd

mqtt:
  use_mqtt: False                         # default=False. Set True true,
to enable mqtt
  server: "your_mqtt_broker"              # No defaults for server
  port: 8883                              # default=8883 for tls, 1883 else
(tls must be "" then !!!!!)
  login: "name"                           # your mqtt user
  password: "your_password"               # password for mqtt user
  tls: "/path/to/your/ca.crt"             # filename including path to your
ca.crt. If not used, must be set to "" !!!!
  device_id: "picframe"                   # default="picframe" unique id of
device. change if there is more than one PictureFrame

http:
  use_http: False                         # default=False. Set True to
enable http NB THIS SERVER IS FOR LOCAL NETWORK AND SHOULD NOT BE EXPOSED
TO EXTERNAL ACCESS
  path: "/home/pi/picframe_data/html"            # path to where html files
are located
  port: 9000                              # port used to serve pages by
http server < 1024 requires root which is *bad* idea
  use_ssl: False
  keyfile: "path/to/key.pem"              # private-key
  certfile: "path/to/cert.pem"            # server certificate

**Comment by jgodfrey** [comment]
Can you attach the file itself rather than pasting its contents? The above paste may make it hard to find a subtle syntax issue...

**Comment by karreglover** [comment]
[configuration.zip](https://github.com/helgeerbe/picframe/files/7834185/configuration.zip)

**Comment by jgodfrey** [comment]
The obvious problem I see is on line 69, here:

`  geo_key: ***@***.***_changed" `

That geo_key is a string, and needs to be surrounded by quotes. Notice that the leading quote is missing. So, it should be something like this:

`  geo_key: "***@***.***_changed" `

**Comment by karreglover** [comment]
I'm not seeing that on my end.  It has both quotes.

geo_key: ***@***.***_changed"

**Comment by jgodfrey** [comment, marked as answer]
Hmmm... Yeah, apologies. I'm not seeing here either now (not sure why I had it to begin with). Anyway, the problem seems to be with the very first line in the file. Specifically, this:

`   viewer:`

That has a leading space (before the `v`) that should not be there. Remove that leading space.

**Comment by karreglover** [comment]
That did it!  Thank you!
